src/environment.rs

```diff
@@ -0,0 +1,129 @@
+        let value_map = HashTrieMap::default();
+        let func_ty: Ty = tcx.type_of(def_id);
+        match func_ty.sty {
+            ty::FnDef(..) | ty::FnPtr(_) => {
+                let sig = func_ty.fn_sig(tcx);
+                let len = sig.inputs().skip_binder().len();
+                for i in 0..len {
+                    let par_i = Path::Parameter { index: i };
+                    // todo: figure out how to get a source span for each parameter definition
+                    value_map.insert(par_i, abstract_value::TOP);
+                }
+            }
+            _ => {}
+        };
+        Environment { value_map }
+    }
+}
+
+/// Methods
+impl Environment {
+    /// Returns a reference to the value associated with the given path.
+    /// If no such value exists, &abstract_value::Bottom is returned.
+    pub fn value_at(&self, path: &Path) -> &AbstractValue {
+        self.value_map.get(path).unwrap_or(&abstract_value::BOTTOM)
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the join of self.value_at(path) and other.value_at(path)
+    pub fn join(&self, other: &Environment, join_condition: &AbstractValue) -> Environment {
+        self.join_or_widen(other, join_condition, |x, y, c| x.join(y, c))
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the widen of self.value_at(path) and other.value_at(path)
+    pub fn widen(&self, other: &Environment, join_condition: &AbstractValue) -> Environment {
+        self.join_or_widen(other, join_condition, |x, y, c| x.widen(y, c))
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the join or widen of self.value_at(path) and other.value_at(path).
+    fn join_or_widen(
+        &self,
+        other: &Environment,
+        join_condition: &AbstractValue,
+        join_or_widen: fn(&AbstractValue, &AbstractValue, &AbstractValue) -> AbstractValue,
+    ) -> Environment {
+        let value_map1 = &self.value_map;
+        let value_map2 = &other.value_map;
+        let mut value_map: HashTrieMap<Path, AbstractValue> = HashTrieMap::default();
+        for (path, val1) in value_map1.iter() {
+            let p = path.clone();
+            match value_map2.get(path) {
+                Some(val2) => {
+                    value_map = value_map.insert(p, join_or_widen(&val1, &val2, &join_condition));
+                }
+                None => {
+                    debug_assert!(!val1.is_bottom());
+                    value_map = value_map.insert(
+                        p,
+                        join_or_widen(&val1, &abstract_value::BOTTOM, &join_condition),
+                    );
+                }
+            }
+        }
+        for (path, val2) in value_map2.iter() {
+            match value_map1.get(path) {
```

Comment:
if !value_map1.contains_key(path) is a little clearer to me, but whatever floats your boat.